drivers/bus/ancillary.c

```diff
@@ -0,0 +1,190 @@
+// SPDX-License-Identifier: GPL-2.0-only
+/*
+ * Software based bus for Ancillary devices
+ *
+ * Copyright (c) 2019-2020 Intel Corporation
+ *
+ * Please see Documentation/driver-api/ancillary_bus.rst for more information.
+ */
+
+#include <linux/device.h>
+#include <linux/init.h>
+#include <linux/module.h>
+#include <linux/pm_domain.h>
+#include <linux/pm_runtime.h>
+#include <linux/string.h>
+#include <linux/ancillary_bus.h>
+
+static bool ancillary_match_id(const struct ancillary_device_id *id,
+			       const struct ancillary_device *adev)
+{
+	while (id->name[0]) {
+		if (!strcmp(adev->match_name, id->name))
+			return true;
+		id++;
+	}
+	return false;
+}
+
+static int ancillary_match(struct device *dev, struct device_driver *drv)
+{
+	struct ancillary_driver *adrv = to_ancillary_drv(drv);
+	struct ancillary_device *adev = to_ancillary_dev(dev);
+
+	return ancillary_match_id(adrv->id_table, adev);
+}
+
+static int ancillary_uevent(struct device *dev, struct kobj_uevent_env *env)
+{
+	struct ancillary_device *adev = to_ancillary_dev(dev);
+
+	if (add_uevent_var(env, "MODALIAS=%s%s", ANCILLARY_MODULE_PREFIX, adev->match_name))
+		return -ENOMEM;
+
+	return 0;
+}
+
+static const struct dev_pm_ops ancillary_dev_pm_ops = {
+	SET_RUNTIME_PM_OPS(pm_generic_runtime_suspend, pm_generic_runtime_resume, NULL)
+	SET_SYSTEM_SLEEP_PM_OPS(pm_generic_suspend, pm_generic_resume)
+};
+
+struct bus_type ancillary_bus_type = {
+	.name = "ancillary",
+	.match = ancillary_match,
+	.uevent = ancillary_uevent,
+	.pm = &ancillary_dev_pm_ops,
+};
+
+/**
+ * ancillary_device_register - add an ancillary bus device
+ * @adev: ancillary bus device to add to the bus
+ * @modname: name of the parent device's driver module
+ */
+int __ancillary_device_register(struct ancillary_device *adev, const char *modname)
+{
+	struct device *dev = &adev->dev;
+	int ret;
+
+	if (!(dev->type && dev->type->release) && !dev->release) {
+		dev_err(dev, "release callback not set for ancillary's device!\n");
+		return -EINVAL;
+	}
+
+	if (!dev->parent) {
+		dev_err(dev, "parent device not set for ancillary's dev!\n");
+		return -EINVAL;
+	}
+
+	/* All error paths out of this function after the device_initialize must perform a
+	 * put_device() so that the .release() callback is called for an error condition.
+	 */
+	device_initialize(dev);
+	dev->bus = &ancillary_bus_type;
+
+	adev->match_name = kasprintf(GFP_KERNEL, "%s.%s", modname, adev->name);
+	if (!adev->match_name) {
+		dev_err(dev, "failed to allocate memory for match_name\n");
+		goto err;
+	}
+
+	ret = dev_set_name(dev, "%s.%d", adev->match_name, adev->id);
+	if (ret) {
+		dev_err(dev, "dev_set_name failed for device\n");
```

Comment:
@dmertman please print the ret code, that keeps it consistent with rest of the code that prints useful error code.

Reply:
done